examples/hetzner-linux/main.tf

```diff
@@ -0,0 +1,117 @@
+terraform {
+  required_providers {
+    coder = {
+      source  = "coder/coder"
+      version = "0.4.1"
+    }
+    hcloud = {
+      source  = "hetznercloud/hcloud"
+      version = "1.33.2"
+    }
+  }
+}
+
+provider "hcloud" {
+  token = var.hcloud_token
+}
+
+provider "coder" {
+}
+
+variable "hcloud_token" {
+  description = <<EOF
+Coder requires a Hetzner Cloud token to provision workspaces.
+EOF
+  sensitive   = true
+}
+
+variable "instance_location" {
+  description = "What region should your workspace live in?"
+  default     = "nbg1"
+  validation {
+    condition     = contains(["nbg1", "fsn1", "hel1"], var.instance_location)
+    error_message = "Invalid zone!"
+  }
+}
+
+variable "instance_type" {
+  description = "What instance type should your workspace use?"
+  default     = "cx11"
+  validation {
+    condition     = contains(["cx11", "cx21", "cx31", "cx41", "cx51"], var.instance_type)
+    error_message = "Invalid instance type!"
+  }
+}
+
+variable "instance_os" {
+  description = "Which operating system should your workspace use?"
+  default     = "ubuntu-20.04"
+  validation {
+    condition     = contains(["ubuntu-22.04","ubuntu-20.04", "ubuntu-18.04", "debian-11", "debian-10", "fedora-35"], var.instance_os)
+    error_message = "Invalid OS!"
+  }
+}
+
+variable "volume_size" {
+  description = "How much storage space do you need?"
+  default     = "50"
+  validation {
+    condition     = contains(["50","100","150"], var.volume_size)
+    error_message = "Invalid volume size!"
+  }
+}
+
+data "coder_workspace" "me" {
+}
+
+resource "coder_agent" "dev" {
+  arch = "amd64"
+  os   = "linux"
+}
+
+resource "hcloud_server" "root" {
+  count = data.coder_workspace.me.start_count
+  name        = "coder-${data.coder_workspace.me.owner}-${data.coder_workspace.me.name}-root"
+  server_type = var.instance_type
+  location    = var.instance_location
+  image       = var.instance_os
+  user_data   = <<EOF
+#!/bin/bash
+mkdir /home/coder/
+mount -o discard,defaults /dev/disk/by-id/scsi-0HC_Volume_${hcloud_volume.root.id} /home/coder/
+export CODER_AGENT_TOKEN=${coder_agent.dev.token}
+${coder_agent.dev.init_script}
+EOF
```

Comment:
I like how nice and simple this setup is! In case you want to add more advanced configuration, you could take a look at the [`do-linux` example](https://github.com/coder/coder/blob/6850db2a47834d0c50142c9e1004195e53c96209/examples/templates/do-linux/main.tf#L126-L131), there we used a cloud-init configuration which, I think, should work here as well.

Reply:
Yes, I think it's fine as is, just wanted to inform you of the option! One use-case I had with the DO droplet was to restart the agent if the droplet rebooted (for whatever reason), do you know if the above script will re-run in such a scenario?

Reply:
I just checked and the coder agent is not restarted on reboot. So I implemented your cloud init template :)